scripts/us_census/acs5yr/subject_tables/common/generate_col_map.py

```diff
@@ -0,0 +1,410 @@
+
+
+class GenerateColMapBase:
+    """module to generate a column map given a list of columns of the dataset and a JSON Spec
+
+  Attributes:
+    specDict: A dictionary containing specifications for the different properties of the statistical variable.
+    columnList: A list of column names for which the column map needs to be generated. This is typically the column header in the dataset.
+    delimiter: The delimiting string that is used for tokenising the column name
+  """
+
+    # TODO - add typing hints and document the format.
+    def __init__(self, spec_dict={}, column_list=[], delimiter='!!'):
+        """module init"""
+        self.features = spec_dict
+        self.column_list = column_list
+        self.column_map = {}
+        self.delimiter = delimiter
+        # TODO: Add a dictionary of counters with summary stats of number of columns, number of statvars generated, statvars of differnt population types, errors or warnings,
+        # fill missing keys in JSON spec with empty values
+        for key in _JSON_KEYS:
+            if key not in self.features:
+                # TODO: initialize UniversePVs as empty list
+                if key == 'ignoreColumns':
+                    self.features[key] = []
+                else:
+                    self.features[key] = {}
+
+    def _find_and_replace_column_names(self, column):
+        """if spec has find_and_replace defined, this function updates column names"""
+        if 'find_and_replace' in self.features['preprocess']:
+            find_and_replace_dict = self.features['preprocess'][
+                'find_and_replace']
+            # replace entire column name
+            if column in find_and_replace_dict:
+                return find_and_replace_dict[column]
+            # replace a token in the column name
+            else:
+                part_list = column.split(self.delimiter)
+                for idx, part in enumerate(part_list):
+                    if part in find_and_replace_dict:
+                        part_list[idx] = find_and_replace_dict[part]
+                        return self.delimiter.join(part_list)
+        return column
+
+    def _generate_stat_vars_from_spec(self):
+        """generates stat_var nodes for each column in column list and returns a dictionary of columns with their respective stat_var nodes
+            Example output: {
+          "Total Civilian population": {
+            "populationType": "Person",
+            "statType": "measuredValue",
+            "measuredProperty": "Count Person"
+            "armedForceStatus": "Civilian"
+          },
+          "<column-name-2>": {}, .....
+        }"""
+        # for each column generate the definition of their respective statistical variable node
+        for col in self.column_list:
+            ignore_token_count = 0
+            for part in col.split(self.delimiter):
+                for token in self.features['ignoreColumns']:
```

Comment:
can be simplified in the future as:
if col in self.features['ignoreColumns'] or
  len((set(self.features['ignoreColumns']) & set(col.split(self.delimiter)) > 0:
   ignore_column = true...

Reply:
Marked as TODO